README.md
# bro-blackbook

### About

`bro-awk` is a way to parse and filter Bro Network Monitor logs with the ability to filter
the columns by the name of the Bro field, rather than the number of the column. This makes
it much more intuitive to use.

`bro-awk` also uses a parallel pipeline to split string processing across all available
processors, making it much faster at crunching through large gzipped logs.

### Installation

This doesn't work yet, but you will eventually be able to install with:

	go get github.com/compilewithstyle/bro-awk

__TODO__

### Usage

__TODO__


CHANGELOG.txt
# DONE #

[2015-02-26] [nsiow] started writing program, laying down basic parallel pipeline
[2015-02-27] [nsiow] finished qreader.go and started making Filter classes
[2015-02-27] [nsiow] started work on the cmd-line interface
[2015-03-01] [nsiow] fleshed out NewFilter function and creation of filters is working well
[2015-03-01] [nsiow] normal string-based searches are working!
[2015-03-04] [nsiow] created the BaseFilter interface that allows for normal/regex filters to be treated agnostically
[2015-03-06] [nsiow] allowed specified fields to be printed, got REGEX filters working

# TODO #

- allow command-line interface to parse flags after other arguments
- allow size of channel / number of workers to be flexible based on the runtime performance of the program


bro-awk.go
/*
	Author:
		Nicholas Siow | compilewithstyle@gmail.com

	Description:
		Implements a fast awk-esque searching through Bro log files
		that allows you to filter the results by statements ... ?
*/

package main

import (
	"bro-awk/qreader"
	"flag"
	"fmt"
	"os"
	"regexp"
)

/*
	Prints a detail usage message showing how the script should be used
*/
func usage() {
	fmt.Println("USAGE:\n\tbro-awk [OPTIONS...] [FILTERS...] [LOGS...]\n")
	fmt.Println("OPTIONS:\n\t-d, --debug\t\tturn on program debugging")
	fmt.Println("\t-p, --print_fields\tonly print the listed fields\n")
	fmt.Println("FILTER SYNTAX:\n\t[literal strings]\n\t<FIELD>=<VALUE>\n\t<FIELD>!=<VALUE>\n\n\t[regexes]\n\t<FIELD>~<VALUE>\n\t<FIELD!~<VALUE>>\n")
	fmt.Println("EXAMPLES:\n\tTODO\n")
	os.Exit(1)
}

/*
	Parses user arguments so that they don't have to use command-line
	flags because those are so 1990s
*/

var log_re *regexp.Regexp = regexp.MustCompile(`.*\.log(?:\.gz)?$`)
var filter_re *regexp.Regexp = regexp.MustCompile(`^\S+(?:=|!=|~|!~)\S+$`)

func parse_args(args []string) ([]string, []string) {

	// make sure at least some arguments were supplied
	if len(args) == 1 {
		fmt.Println("[ERROR] not enough arguments")
		os.Exit(1)
	}

	// if not, then continue to parse the arguments, adding them
	// to the appropriate slices
	logs := make([]string, 0)
	filters := make([]string, 0)

	for _, arg := range args[1:] {
		if filter_re.MatchString(arg) {
			filters = append(filters, arg)
		} else if log_re.MatchString(arg) {
			logs = append(logs, arg)
		}
	}

	// make sure that some parameters were supplied for both logs and filters

	if len(logs) == 0 {
		fmt.Println("[ERROR] No logs specified. Use `bro-awk --help` for more info")
		os.Exit(1)
	}

	if len(filters) == 0 {
		fmt.Println("[ERROR] No filters specified. Use `bro-awk --help` for more info")
		os.Exit(1)
	}

	return logs, filters
}

/*
	Using the given arguments, construct the necessary filters and run them against the logs
*/
func main() {
	// first, check to see if usage string was requested
	for _, x := range os.Args {
		if x == "-h" || x == "--help" {
			usage()
		}
	}

	// next, parse the option flags
	print_fields := flag.String("p", "", "")
	flag.Parse()

	// next, parse through the remaining arguments to find user-supplied filters and logs
	logs, filters := parse_args(os.Args)

	// create a new Qreader:
	// 		unzipper, []string of filters, number of processors, reading blocksize
	q := qreader.NewQreader("", filters, 0, 0, *print_fields)

	// iterate through the logs and apply the filter to each of them
	for _, log := range logs {
		q.Parse(log)
	}
}


filters/filters.go
package filters

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

/*
	global indexmap which allows mapping from field -> index in Linedata slice

	this allows us to pass around []string instead of map[string]string and just
	use this map to index into the given field
*/
var indexmap map[string]int

//--------------------------------------------------------------------------------
//	Linedata wrapper for []string
//--------------------------------------------------------------------------------

/* define a custom wrapper for []string that allows helper functions */
type Linedata []string

/*
	helper function that allows for easy indexing into a Linedata struct
	via the name of the field you're interested in
*/
func (self Linedata) get(field string) string {
	idx, ok := indexmap[field]
	if !ok {
		fmt.Printf("[ERROR] unable to find index for field: %s\n", field)
		fmt.Println("indexmap dump:")
		fmt.Println(indexmap)
		os.Exit(1)
	}

	return self[idx]
}

//--------------------------------------------------------------------------------
//	Single Filter class
//--------------------------------------------------------------------------------

/*
	Filter interface that allows agnostic treatment of string/regex based filters
*/
type BaseFilter interface {
	Passes(data *Linedata) bool
}

/*
	Filter struct that represents a single, non-regex rule
*/
type Filter struct {
	fields           []string
	values           []string
	compare_function func(a string, b string) bool
}

/*
	Filter struct that represents a single, regex-based rule
*/
type RegexFilter struct {
	fields           []string
	values           []*regexp.Regexp
	compare_function func(a string, re *regexp.Regexp) bool
}

/*
	Constructor for single filter type
	TODO check with regex or something to make sure it's a valid rule!
*/
func NewFilter(rule string) BaseFilter {
	// set the appropriate comparison function based on which
	// operator is given
	var op string
	var isregex, negate bool

	if strings.Contains(rule, "!=") {
		op = "!="
		negate = true
		isregex = false
	} else if strings.Contains(rule, "=") {
		op = "="
		negate = false
		isregex = false
	} else if strings.Contains(rule, "!~") {
		op = "!~"
		negate = true
		isregex = true
	} else if strings.Contains(rule, "~") {
		op = "~"
		negate = false
		isregex = true
	} else {
		fmt.Println("[ERROR] not sure how to parse rule: " + rule)
		os.Exit(1)
	}

	// split the rule into fields/values and set the appropriate fields
	// in the new filter
	opsides := strings.Split(rule, op)
	if len(opsides) != 2 {
		fmt.Println("[ERROR] rule contains too many boolean operators: " + rule)
		os.Exit(1)
	}

	fields := strings.Split(opsides[0], ",")
	values := strings.Split(opsides[1], ",")

	// choose the comparison operator based on whether or not to negate
	// the filter
	if isregex {
		f := &RegexFilter{}

		// compile the user-supplied regexes
		regex_values := make([]*regexp.Regexp, len(values))
		for i, v := range values {
			my_regex, err := regexp.Compile(v)
			if err != nil {
				fmt.Println("[ERROR] unable to compile POSIX regex: " + v)
				os.Exit(1)
			} else {
				regex_values[i] = my_regex
			}
		}

		// set the fields and values of the filter
		f.fields = fields
		f.values = regex_values

		// set the compare function based on whether or not negation should be used
		if negate {
			f.compare_function = func(a string, re *regexp.Regexp) bool {
				return !re.MatchString(a)
			}
		} else {
			f.compare_function = func(a string, re *regexp.Regexp) bool {
				return re.MatchString(a)
			}
		}

		return BaseFilter(f)
	} else {
		f := &Filter{}

		// set the fields and values of the filter
		f.fields = fields
		f.values = values

		// set the compare function based on whether or not negation should be used
		if negate {
			f.compare_function = func(a string, b string) bool {
				return (a != b)
			}
		} else {
			f.compare_function = func(a string, b string) bool {
				return (a == b)
			}
		}

		return BaseFilter(f)
	}
}

/*
	Determines whether or not that line passes based off the given filter
	TODO
*/
func (self Filter) Passes(data *Linedata) bool {
	for _, field := range self.fields {
		for _, value := range self.values {
			if self.compare_function(data.get(field), value) {
				return true
			}
		}
	}

	return false
}

/*
	Determines whether or not that line passes based off the given filter
	TODO
*/
func (self RegexFilter) Passes(data *Linedata) bool {
	for _, field := range self.fields {
		for _, value := range self.values {
			if self.compare_function(data.get(field), value) {
				return true
			}
		}
	}

	return false
}

//--------------------------------------------------------------------------------
//	Aggregate FilterSet class
//--------------------------------------------------------------------------------

/*
	Combined set of filters that can be used to match against a given line

	The line must match all contained filters in order for it to 'pass'
*/
type FilterSet struct {
	filters []BaseFilter
}

/*
	Helper function to set up the index map and turn string parameters
	into Filter objects
*/
func NewFilterSet(params []string) *FilterSet {
	fs := FilterSet{}
	fs.filters = make([]BaseFilter, len(params))

	for i, param_string := range params {
		fs.filters[i] = NewFilter(param_string)
	}

	return &fs
}

/*
	Function that creates the indexmap for these filters using the Bro
	header for a given file
*/
func (self FilterSet) ApplyHeader(header []string) {
	indexmap = make(map[string]int)

	for idx, field := range header {
		indexmap[field] = idx
	}
}

/*
	Official interface for Filter program -- test lines against
	this to determine whether or not they should be printed
*/
func (self FilterSet) Passes(data *Linedata) bool {
	for _, f := range self.filters {
		if !f.Passes(data) {
			return false
		}
	}

	return true
}


logreader/logreader.go
/*
	Author:
		Nicholas Siow | compilewithstyle@gmail.com

	Description:
		Super-fast gzip reader that uses goroutine pools to
		eliminate bottlenecks in filereading
*/

package logreader

import (
	"../filters"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

//--------------------------------------------------------------------------------
//	PROGRAM VARIABLES
//--------------------------------------------------------------------------------

var chansize int = 10000

//--------------------------------------------------------------------------------
//	READER
//--------------------------------------------------------------------------------

/*
	Reader class which handles the
*/
type Reader struct {
	filename string
	unzipper string
	bsize    int
	outq     chan []byte
}

/*
	Returns an appropriate io.Reader object based on whether or not
	the file is gzipped. Uses the `Unzipper` variable to determine
	which program to use in the case of a gzipped file
*/
func (self Reader) GetReader() io.Reader {
	if strings.HasSuffix(self.filename, ".gz") {

		// init a subprocess using the Unzipper command
		// TODO -- let the -c be an option
		c := exec.Command(self.unzipper, "-c", self.filename)
		pipe, err := c.StdoutPipe()
		if err != nil {
			panic(err)
		}

		// start the subprocess and return a reader connected to
		// its STDOUT
		c.Start()
		return pipe

	} else {

		// otherwise just open a file as normal and return
		// an io.Reader object for it
		file, err := os.Open(self.filename)
		if err != nil {
			panic(err)
		}

		return io.Reader(file)

	}
}

/*
	Begins to read from the given file and pushes data
	into a channel. Closes the channel upon EOF
*/
func (self Reader) Start() {
	// get an appropriate reader
	reader := self.GetReader()

	// initialize a byteslice for the partial lines
	// to be added to the following read chunk
	var leftovers []byte

	// loop until EOF
	for {
		// read in the next chunk
		buffer := make([]byte, self.bsize)
		length, err := reader.Read(buffer)
		if err != nil && err != io.EOF {
			panic(err)
		}

		// break if reading is done
		if length == 0 {
			break
		}

		// add partial line from previous chunk to beginning of this chunk
		// add partial line from this chunk to leftovers variable for next
		end_it := length - 1
		for {
			if end_it == 0 {
				leftovers = append(leftovers, buffer...)
				break
			}
			if buffer[end_it] == '\n' {
				leftovers = append(leftovers, buffer[:end_it]...)
				self.outq <- leftovers
				leftovers = buffer[end_it+1:]
				break
			} else {
				end_it -= 1
			}
		}
	}

	// close channel to let next worker know that you're done
	close(self.outq)
}

//--------------------------------------------------------------------------------
//	PARSER
//--------------------------------------------------------------------------------

/*
	Parser class which handles splitting data at newlines and separating
	out relevant data
*/
type Parser struct {
	filter  *filters.FilterSet
	limiter chan int
	inq     chan []byte
}

func (self Parser) Parse(fileslice []byte) {
	// split incoming byteslice @ newlines
	raw_lines := strings.Split(string(fileslice), "\n")

	for _, line := range raw_lines {
		// skip commented lines
		if line[0] == '#' {
			continue
		}

		// split on tabs to create Linedata object
		var ld filters.Linedata = strings.Split(line, "\t")
		if self.filter.Passes(&ld) {
			fmt.Println(line)
		}
	}

	<-self.limiter
}

func (self Parser) Start() {
	for fileslice := range self.inq {
		self.limiter <- 1
		go self.Parse(fileslice)
	}

	for {
		if len(self.limiter) == 0 {
			break
		} else {
			time.Sleep(500 * time.Millisecond)
		}
	}
}

//--------------------------------------------------------------------------------
//	MAIN QREADER CLASS
//--------------------------------------------------------------------------------

type Qreader struct {
	Filename   string
	Unzipper   string
	ParserPool int
	Blocksize  int
	Filter     *filters.FilterSet
}

/*
	Struct initializer for QREADER
*/
func NewQreader(Unzipper string, filter_strings []string, ParserPool int, Blocksize int) *Qreader {
	// initialize a new, empty Qreader
	q := Qreader{}

	// set the unzipper, find one if not given
	if Unzipper == "" {
		Unzipper = FindUnzipper()
	}
	q.Unzipper = Unzipper

	// set the number of workers in the parser pool, use default if not given
	if ParserPool <= 0 {
		ParserPool = runtime.NumCPU() - 1
	}
	q.ParserPool = ParserPool

	// set the reading blocksize, use default if not given
	if Blocksize <= 0 {
		Blocksize = 8192
	}
	q.Blocksize = Blocksize

	// set the number of max concurrent goroutines
	runtime.GOMAXPROCS(runtime.NumCPU() - 1)

	// set up the filters
	q.Filter = filters.NewFilterSet(filter_strings)

	return &q
}

/*
	Function to find a program for gz decompression
*/
func FindUnzipper() string {
	possibilities := []string{"unpigz", "gzcat", "zcat"}

	for _, p := range possibilities {
		cmd, err := exec.LookPath(p)
		if err == nil {
			return cmd
		}
	}

	fmt.Println("[ERROR] could not find a program for gz decompression")
	os.Exit(1)
	return ""
}

/*
	Read in the bro log file up to the `#fields` line and find the names of the various fields
*/
func GetHeader(unzipper string, fn string) []string {
	cmdstring := fmt.Sprintf("%s -c %s | grep -m1 fields", unzipper, fn)
	cmd := exec.Command("bash", "-c", cmdstring)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		panic(err)
	}

	cmd.Start()

	field_string, err := ioutil.ReadAll(stdout)
	if err != nil {
		panic(err)
	}

	// to from 0:end-1 of the field string (to ignore newline)
	// then take 1:end to get everything except the '#fields' string
	return strings.Split(string(field_string[:len(field_string)-1]), "\t")[1:]
}

/*
	Set up the workers and read through a given file
*/
func (self Qreader) Parse(fn string) {
	// find the header for the bro file
	header := GetHeader(self.Unzipper, fn)

	// use the header and the filter strings to generate a FilterSet
	// TODO find a more elegant way of doing this??
	self.Filter.ApplyHeader(header)

	// create the necessary channels
	chan1 := make(chan []byte, chansize)

	// create buffered controller channels that can act as semaphores
	// to limit overall throughput
	limiter1 := make(chan int, self.ParserPool)

	// intialize the various worker objects
	r := Reader{fn, self.Unzipper, self.Blocksize, chan1}
	p := Parser{self.Filter, limiter1, chan1}

	// start each of the worker functions on its own goroutine
	go r.Start()
	p.Start()
}


qreader/qreader.go
/*
	Author:
		Nicholas Siow | compilewithstyle@gmail.com

	Description:
		Super-fast gzip reader that uses goroutine pools to
		eliminate bottlenecks in filereading
*/

package qreader

import (
	"bro-awk/filters"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

//--------------------------------------------------------------------------------
//	PROGRAM VARIABLES
//--------------------------------------------------------------------------------

var chansize int = 10000

//--------------------------------------------------------------------------------
//	READER
//--------------------------------------------------------------------------------

/*
	Reader class which handles the
*/
type Reader struct {
	filename string
	unzipper string
	bsize    int
	outq     chan []byte
}

/*
	Returns an appropriate io.Reader object based on whether or not
	the file is gzipped. Uses the `Unzipper` variable to determine
	which program to use in the case of a gzipped file
*/
func (self Reader) GetReader() io.Reader {
	if strings.HasSuffix(self.filename, ".gz") {

		// init a subprocess using the Unzipper command
		// TODO -- let the -c be an option
		c := exec.Command(self.unzipper, "-c", self.filename)
		pipe, err := c.StdoutPipe()
		if err != nil {
			panic(err)
		}

		// start the subprocess and return a reader connected to
		// its STDOUT
		c.Start()
		return pipe

	} else {

		// otherwise just open a file as normal and return
		// an io.Reader object for it
		file, err := os.Open(self.filename)
		if err != nil {
			panic(err)
		}

		return io.Reader(file)

	}
}

/*
	Begins to read from the given file and pushes data
	into a channel. Closes the channel upon EOF
*/
func (self Reader) Start() {
	// get an appropriate reader
	reader := self.GetReader()

	// initialize a byteslice for the partial lines
	// to be added to the following read chunk
	var leftovers []byte

	// loop until EOF
	for {
		// read in the next chunk
		buffer := make([]byte, self.bsize)
		length, err := reader.Read(buffer)
		if err != nil && err != io.EOF {
			panic(err)
		}

		// break if reading is done
		if length == 0 {
			break
		}

		// add partial line from previous chunk to beginning of this chunk
		// add partial line from this chunk to leftovers variable for next
		end_it := length - 1
		for {
			if end_it == 0 {
				leftovers = append(leftovers, buffer...)
				break
			}
			if buffer[end_it] == '\n' {
				leftovers = append(leftovers, buffer[:end_it]...)
				self.outq <- leftovers
				leftovers = buffer[end_it+1:]
				break
			} else {
				end_it -= 1
			}
		}
	}

	// close channel to let next worker know that you're done
	close(self.outq)
}

//--------------------------------------------------------------------------------
//	PARSER
//--------------------------------------------------------------------------------

/*
	Parser class which handles splitting data at newlines and separating
	out relevant data
*/
type Parser struct {
	filter          *filters.FilterSet
	limiter         chan int
	inq             chan []byte
	print_indices   []int
	selective_print bool
}

func (self Parser) Parse(fileslice []byte) {
	// split incoming byteslice @ newlines
	raw_lines := strings.Split(string(fileslice), "\n")

	for _, line := range raw_lines {
		// skip commented lines
		if line[0] == '#' {
			continue
		}

		// split on tabs to create Linedata object
		var ld filters.Linedata = strings.Split(line, "\t")
		if self.filter.Passes(&ld) {
			// print the specified fields, or the whole line if none were specifically asked for
			if self.selective_print {
				to_print := make([]string, len(self.print_indices))
				for i, idx := range self.print_indices {
					to_print[i] = ld[idx]
				}

				fmt.Println(strings.Join(to_print, "\t"))
			} else {
				fmt.Println(line)
			}
		}
	}

	<-self.limiter
}

func (self Parser) Start() {
	for fileslice := range self.inq {
		self.limiter <- 1
		go self.Parse(fileslice)
	}

	for {
		if len(self.limiter) == 0 {
			break
		} else {
			time.Sleep(500 * time.Millisecond)
		}
	}
}

//--------------------------------------------------------------------------------
//	MAIN QREADER CLASS
//--------------------------------------------------------------------------------

type Qreader struct {
	Filename       string
	Unzipper       string
	ParserPool     int
	Blocksize      int
	Filter         *filters.FilterSet
	PrintFields    []string
	PrintIndices   []int
	SelectivePrint bool
}

/*
	Struct initializer for QREADER
*/
func NewQreader(Unzipper string, filter_strings []string, ParserPool int, Blocksize int, my_print_fields string) *Qreader {
	// initialize a new, empty Qreader
	q := Qreader{}

	// set the unzipper, find one if not given
	if Unzipper == "" {
		Unzipper = FindUnzipper()
	}
	q.Unzipper = Unzipper

	// set the number of workers in the parser pool, use default if not given
	if ParserPool <= 0 {
		ParserPool = runtime.NumCPU() - 1
	}
	q.ParserPool = ParserPool

	// set the reading blocksize, use default if not given
	if Blocksize <= 0 {
		Blocksize = 8192
	}
	q.Blocksize = Blocksize

	// set the number of max concurrent goroutines
	runtime.GOMAXPROCS(runtime.NumCPU() - 1)

	// set up the filters
	q.Filter = filters.NewFilterSet(filter_strings)

	// if print_fields is given, set that global variable
	if my_print_fields == "" {
		q.PrintFields = nil
		q.SelectivePrint = false
	} else {
		q.PrintFields = strings.Split(my_print_fields, ",")
		q.SelectivePrint = true
	}

	return &q
}

/*
	Function to find a program for gz decompression
*/
func FindUnzipper() string {
	possibilities := []string{"gzcat", "unpigz", "zcat"}

	for _, p := range possibilities {
		cmd, err := exec.LookPath(p)
		if err == nil {
			return cmd
		}
	}

	fmt.Println("[ERROR] could not find a program for gz decompression")
	os.Exit(1)
	return ""
}

/*
	Read in the bro log file up to the `#fields` line and find the names of the various fields
*/
func GetHeader(unzipper string, fn string) []string {
	cmdstring := fmt.Sprintf("%s -c %s | grep -m1 fields", unzipper, fn)
	cmd := exec.Command("bash", "-c", cmdstring)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		panic(err)
	}

	cmd.Start()

	field_string, err := ioutil.ReadAll(stdout)
	if err != nil {
		panic(err)
	}

	// to from 0:end-1 of the field string (to ignore newline)
	// then take 1:end to get everything except the '#fields' string
	return strings.Split(string(field_string[:len(field_string)-1]), "\t")[1:]
}

/*
	Set up the workers and read through a given file
*/
func (self Qreader) Parse(fn string) {
	// find the header for the bro file
	header := GetHeader(self.Unzipper, fn)

	// if only certain fields are to be printed, use the new header to determine
	// the indices of those fields
	if self.SelectivePrint {
		self.PrintIndices = make([]int, len(self.PrintFields))

		for i1, field := range self.PrintFields {
			for i2, header_field := range header {
				if field == header_field {
					self.PrintIndices[i1] = i2
				}
			}
		}
	} else {
		self.PrintIndices = nil
	}

	// use the header and the filter strings to generate a FilterSet
	// TODO find a more elegant way of doing this??
	self.Filter.ApplyHeader(header)

	// create the necessary channels
	chan1 := make(chan []byte, chansize)

	// create buffered controller channels that can act as semaphores
	// to limit overall throughput
	limiter1 := make(chan int, self.ParserPool)

	// intialize the various worker objects
	r := Reader{fn, self.Unzipper, self.Blocksize, chan1}
	p := Parser{self.Filter, limiter1, chan1, self.PrintIndices, self.SelectivePrint}

	// start each of the worker functions on its own goroutine
	go r.Start()
	p.Start()
}
